### PR TITLE
Develop profiler

### PR DESCRIPTION
**Pull request by eshvk**
@Dominique-Unity
@rejunit
@mantasp

### CODE REVIEW

**Comment by vincentpierre** [comment]
Can you update https://github.com/Unity-Technologies/ml-agents/blob/develop/docs/Python-API.md ?

File: python/learn.py [hunk 1]
```diff
@@ -46,6 +46,7 @@
       --train                    Whether to train model, or only run inference [default: False].
       --worker-id=<n>            Number to add to communication port (5005). Used for multi-environment [default: 0].
       --docker-target-name=<dt>       Docker Volume to store curriculum, executable and model files [default: Empty].
+      --no-graphics                     Whether to run the Unity simulator in no-graphics mode [default: False].
```
Comment (vincentpierre) [inline python/learn.py]: I think there is an extra tab here

File: python/unitytrainers/trainer.py [hunk 2]
```diff
@@ -76,17 +76,11 @@ def get_last_reward(self):
...
+        Increment the step count of the trainer and updates the last reward
         """
-        raise UnityTrainerException("The increment_step method was not implemented.")
-
-    def update_last_reward(self):
-        """
-        Updates the last reward
-        """
-        raise UnityTrainerException("The update_last_reward method was not implemented.")
+        raise UnityTrainerException("The increment_step_and_update_last_reward method was not implemented.")
```
Comment (vincentpierre) [inline python/unitytrainers/trainer.py]: This needs a comment

File: python/unitytrainers/models.py [hunk 3]
```diff
@@ -85,6 +82,15 @@ def create_vector_input(self, name='vector_observation'):
             self.vector_in = tf.placeholder(shape=[None, 1], dtype=tf.int32, name='vector_observation')
             return self.vector_in
 
+    def create_normalizer_update(self):
```
Comment (vincentpierre) [inline python/unitytrainers/models.py]: I would rather have this method return `update_mean` and `self.update_variance` and take in `self.vector_in` as input.

File: python/unitytrainers/bc/trainer.py [hunk 4]
```diff
@@ -122,16 +122,11 @@ def get_last_reward(self):
         else:
             return 0
 
-    def increment_step(self):
+    def increment_step_and_update_last_reward(self):
```
Comment (vincentpierre) [inline python/unitytrainers/bc/trainer.py]: What is the reason behind this change ?

File: python/unitytrainers/ppo/trainer.py [hunk 5]
```diff
@@ -95,8 +95,18 @@ def __init__(self, sess, env, brain_name, trainer_parameters, training, seed):
 
         self.summary_writer = tf.summary.FileWriter(self.summary_path)
 
+        self.inference_run_list = [self.model.output, self.model.all_probs, self.model.value,
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/trainer.py]: Can the same run_list system be implemented in BC ?

File: python/unitytrainers/ppo/trainer.py [hunk 6]
```diff
@@ -95,8 +95,18 @@ def __init__(self, sess, env, brain_name, trainer_parameters, training, seed):
...
+                                   self.model.entropy, self.model.learning_rate]
+        if self.is_continuous_action:
+            self.inference_run_list.append(self.model.output_pre)
+        if self.use_recurrent:
+            self.inference_run_list.extend([self.model.memory_out])
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/trainer.py]: Has this been tested out with the hallway ?

File: python/unitytrainers/bc/trainer.py (hunk 4 above)
Comment (awjuliani) [reply python/unitytrainers/bc/trainer.py]: Allows only a single call to the network to get both, as calls to the network are expensive, even for grabbing single variables.

File: python/unitytrainers/ppo/trainer.py (hunk 6 above)
Comment (awjuliani) [reply python/unitytrainers/ppo/trainer.py]: Yep.

File: python/unitytrainers/ppo/trainer.py (hunk 5 above)
Comment (awjuliani) [reply python/unitytrainers/ppo/trainer.py]: Just did it.

File: python/unitytrainers/models.py (hunk 3 above)
Comment (awjuliani) [reply python/unitytrainers/models.py]: Done.

File: python/learn.py (hunk 1 above)
Comment (awjuliani) [reply python/learn.py]: It's gone.

**Comment by vincentpierre** [comment]
🚢 🇮🇹
